File: node_version/services/emailService.js
import formData from 'form-data';
import Mailgun from 'mailgun.js';
import { differenceInMinutes } from 'date-fns';

import { db, timestamp } from '../db.js';
import { decryptText } from './securityService.js';

const mailgun = new Mailgun(formData);

const sendEmail = async (uid, template, batch) => {
    // Fetch the user document from Firestore
    const userDoc = await db.collection('clients').doc(uid).get();
    const userData = userDoc.data();

    // Check if the user has their own Mailgun config, bypass trial logic if true
    if (!userData.hasMailgunConfig) {
        // Check if the user is on trial and has messages left
        if (userData.onTrial && userData.messagesLeft <= 0) {
            throw new Error('Trial has ended, no messages left');
        }
    }

    const encryptedMailgunApiKey = userData['mailgunApiKey'];

    const mailgunApiKey = await decryptText(encryptedMailgunApiKey);
    const mailgunDomain = userData['mailgunDomain'];

    // Initialize the Mailgun client with the API key from the user document
    const client = mailgun.client({
        username: 'api',
        key: mailgunApiKey,
    });

    template.to_email = template.to_email.toLowerCase();

    // Generate a unique document ID
    let docId = db.collection('clients').doc().id;

    const messageData = {
        from: `${template.from_name} <${template.from_email}>`,
        to: `${template.to_name} <${template.to_email}>`,
        subject: template.subject,
        text: template.message,
        html: `<html><body>${template.message}</body></html>`,
        'h:Message-ID': `<${uid}-${docId}@${mailgunDomain}>`,
    };

    try {
        const res = await client.messages.create(mailgunDomain, messageData);
        if (res.status === 200) {
            // If the user is on trial and does not have their own Mailgun config, decrement the messagesLeft field
            if (userData.onTrial && !userData.hasMailgunConfig) {
                await db
                    .collection('clients')
                    .doc(uid)
                    .update({
                        messagesLeft: userData.messagesLeft - 1,
                    });
            }

            let docRef = db
                .collection('clients')
                .doc(uid)
                .collection('emails')
                .doc(docId);
            batch.set(docRef, {
                to_email: template.to_email,
                to_name: template.to_name,
                to_company: template.to_company,
                from_name: template.from_name,
                email: [
                    {
                        role: 'assistant',
                        content: template.message,
                    },
                ],
                sent_timestamp: timestamp,
                follow_up_1_sent: false,
                follow_up_2_sent: false,
                response_received: false,
            });
        }
    } catch (err) {
        console.error(err);
    }
};

export const handleEmailSending = async (data) => {
    let batch = db.batch();
    // Check if email_templates is an array
    if (Array.isArray(data.emailTemplates)) {
        // Send multiple emails
        for (let template of data.emailTemplates) {
            await sendEmail(data.uid, template, batch);
        }
    } else {
        // Send a single to_email
        const requiredFields = [
            'from_name',
            'from_email',
            'to_email',
            'to_name',
            'subject',
            'message',
        ];
        for (let field of requiredFields) {
            if (!data.emailTemplates[field]) {
                return { message: `${field} is required`, status: 400 };
            }
        }
        await sendEmail(data.uid, data.emailTemplates, batch);
    }
    await batch.commit();
};

export const handleFollowUps = async () => {
    let sentEmails = [];
    // Fetch all user documents from Firestore
    const usersSnapshot = await db.collection('clients').get();
    const users = usersSnapshot.docs.map((doc) => ({
        uid: doc.id,
        ...doc.data(),
    }));

    for (let userData of users) {
        const uid = userData.uid;

        // Initialize the Mailgun client with the API key from the user document
        const client = mailgun.client({
            username: 'api',
            key: userData['mailgunApiKey'],
        });

        // Fetch follow-up emails from Firestore
        const followUpEmailsSnapshot = await db
            .collection('clients')
            .doc(uid)
            .collection('follow-up-email-templates')
            .get();
        const followUpEmails = followUpEmailsSnapshot.docs.map((doc) =>
            doc.data()
        );

        // Access the 'emails' subcollection
        const emails = db.collection('clients').doc(uid).collection('emails');
        const snapshot = await emails
            .where('response_received', '==', false)
            .get();
        const docs = snapshot.docs;

        for (let doc of docs) {
            const data = doc.data();
            const timeSinceSent = differenceInMinutes(
                new Date(),
                data.sent_timestamp.toDate()
            );

            if (timeSinceSent > 3 && !data.follow_up_1_sent) {
                const followUp = followUpEmails.find(
                    (email) => email.id === 'follow-up1'
                );

                const messageData = {
                    from: `${userData.fromNames[0]} <${userData.fromNames[0]}@${userData['mailgunDomain']}>`,
                    to: `${data.to_name} <${data.to_email}>`,
                    subject: followUp.subject,
                    text: followUp.message,
                    html: `<html><body>${followUp.message}</body></html>`,
                };

                try {
                    const res = await client.messages.create(
                        userData['mailgun-domain'],
                        messageData
                    );
                    if (res.status === 200) {
                        await doc.ref.update({ follow_up_1_sent: true });
                        sentEmails.push(messageData);
                    }
                } catch (err) {
                    console.error(err);
                }
            } else if (timeSinceSent > 5 && !data.follow_up_2_sent) {
                const followUp = followUpEmails.find(
                    (email) => email.id === 'follow-up2'
                );

                const messageData = {
                    from: `${userData.fromNames[0]} <${userData.fromNames[0]}@${userData['mailgunDomain']}>`,
                    to: `${data.to_name} <${data.to_email}>`,
                    subject: followUp.subject,
                    text: followUp.message,
                    html: `<html><body>${followUp.message}</body></html>`,
                };
                try {
                    const res = await client.messages.create(
                        userData['mailgun-domain'],
                        messageData
                    );
                    if (res.status === 200) {
                        await doc.ref.update({ follow_up_2_sent: true });
                        sentEmails.push(messageData);
                    }
                } catch (err) {
                    console.error(err);
                }
            }
        }
    }
    return sentEmails;
};

File: python_version/services/email_service.py
import os
import requests
from datetime import datetime
from google.api_core.exceptions import GoogleAPICallError
from flask import current_app

api_keys = {
    'testing': os.getenv('DEV_TEST_KEY'),
}

def send_custom_message(data):
    required_fields = ['sender_name', 'sender_email', 'recipient_email', 'recipient_name', 'recipient_company', 'subject', 'message']
    for field in required_fields:
        if not data.get(field):
            return {'message': f'{field} is required', 'status': 400}
        
    user_id = data.get('user_id')
    sender_name = data.get('sender_name')
    sender_email = data.get('sender_email')
    recipient_email = data.get('recipient_email')
    recipient_name = data.get('recipient_name')
    recipient_company = data.get('recipient_company')
    subject = data.get('subject')
    message = data.get('message')

    api_key = api_keys.get(user_id)
    if not api_key:
        return {'message': 'Invalid user ID', 'status': 400}
    
    html_content = f"""
    <html>
        <body>
            {message}      
        </body>
    </html>
    """

    try:
        response = requests.post(
            "https://api.mailgun.net/v3/mg.shauno.co/messages", 
            auth=("api", api_key),
            data={"from": f"{sender_name} <{sender_email}>",
                  "to": [recipient_email],
                  "subject": subject,
                  "text": message,
                  "html": html_content,},
            timeout=5) 
    except requests.exceptions.RequestException as e:
        return {'message': str(e), 'status': 500}

    if response.status_code != 200:
        return {'message': f"Error: {response.status_code}, {response.text}", 'status': response.status_code}

    try:
        doc_ref = current_app.db.collection('clients').document('uid').collection('emails').document()
        doc_ref.set({
            'recipient_email': recipient_email,
            'recipient_name': recipient_name,
            'recipient_company': recipient_company,
            'sender_name': sender_name,
            'sent_timestamp': datetime.now(),
            'follow_up_1_sent': False,
            'follow_up_2_sent': False,
            'response_received': False,
        })
    except (ValueError, TypeError, GoogleAPICallError) as e:
        return {'message': f"Database error: {str(e)}", 'status': 500}

    return {'message': 'Email sent successfully', 'status': 200}

def send_multiple_messages(email_templates, user_id):
    results = []
    for template in email_templates:
        data = {
            'user_id': user_id,
            'sender_name': 'John Doe',
            'sender_email': 'test@mg.shauno.co',
            'recipient_email': template['email'],
            'recipient_name': template['recipient_name'],
            'recipient_company': template['recipient_company'],
            'subject': template['subject'],
            'message': template['message']
        }
        result = send_custom_message(data)
        if result['status'] != 200:
            return result
        results.append(result)
    if not results:
        return {'message': 'No emails were sent', 'status': 400}
    return {'message': 'All emails sent successfully', 'status': 200}

def fetch_emails():
    try:
        docs = current_app.db.collection('clients').document('uid').collection('emails').get()
        emails = [doc.to_dict() for doc in docs]
        return emails
    except (ValueError, TypeError, GoogleAPICallError) as e:
        return {'message': f"Database error: {str(e)}"}, 500